editor/src/core/model/project-import.ts

```diff
@@ -0,0 +1,133 @@
+  const fileBuffer = await file.async('nodebuffer')
+  if (isText(fileName, fileBuffer)) {
+    return file.async('string')
+  } else {
+    return null
+  }
+}
+
+export interface UnsavedAsset {
+  fileName: string
+  fileType: string
+  base64: string
+  hash: string
+  size: Size | null
+}
+
+export interface ProjectImportResult {
+  projectName: string
+  contents: ProjectContents
+  assetsToUpload: Array<UnsavedAsset>
+}
+
+export async function importZippedGitProject(
+  projectName: string,
+  zipped: JSZip,
+): Promise<ProjectImportResult> {
+  let loadedProject: ProjectContents = {}
+  let promises: Array<Promise<void>> = []
+  let assetsToUpload: Array<UnsavedAsset> = []
+
+  const loadFile = async (fileName: string, file: JSZip.JSZipObject) => {
+    const shiftedFileName = fileName.replace(/[^\/]*\//, '/') // Github uses the project name and a commit hash as the root directory
+    if (shiftedFileName.trim() === '/') {
+      // We don't need to create a root directory
+      return
+    }
+    if (file.dir) {
+      loadedProject[shiftedFileName] = directory()
+    } else {
+      const expectedFileType = fileTypeFromFileName(shiftedFileName)
+      switch (expectedFileType) {
+        case 'ASSET_FILE': {
+          const fileType = getFileExtension(shiftedFileName)
+          const base64 = await file.async('base64')
+          const assetResult = assetResultForBase64(shiftedFileName, base64)
+          assetsToUpload.push({
+            fileName: assetResult.filename,
+            fileType: fileType,
+            base64: assetResult.dataUrl,
+            hash: assetResult.hash,
+            size: null,
+          })
+          loadedProject[shiftedFileName] = assetFile()
+          break
+        }
+        case 'IMAGE_FILE': {
+          const fileType = getFileExtension(shiftedFileName)
+          const base64 = await file.async('base64')
+          const imageResult = await imageResultForBase64(shiftedFileName, fileType, base64)
+          assetsToUpload.push({
+            fileName: imageResult.filename,
+            fileType: fileType,
+            base64: imageResult.dataUrl,
+            hash: imageResult.hash,
+            size: imageResult.size,
+          })
+          loadedProject[shiftedFileName] = imageFile(
+            undefined,
+            undefined,
+            undefined,
+            undefined,
+            imageResult.hash,
+          )
+          break
+        }
+        case 'UI_JS_FILE':
+        case 'CODE_FILE':
+          const loadedFile = await attemptedTextFileLoad(shiftedFileName, file)
+          if (loadedFile == null) {
+            // FIXME Client should show an error if loading a text file fails
+            console.error(`Failed to parse file ${shiftedFileName} as a text file`)
+          } else {
+            if (expectedFileType === 'UI_JS_FILE') {
+              // TODO We really need a way to represent unparsed files, or a way to separate the parsed file from the contents
```

Comment:
Subsequent work or a nice to have?

Reply:
Hmmm somewhere in between. Not having that makes the code quite clunky in a number of places.